logging: add syst format support for existing logging system

wentongwu opened the pull request:
add syst format support for existing logging system, the output of sys-t format can be decoded by the existing decoder.

Fixes: #19841.

andrewboie commented on west.yml:
```diff
@@ -21,6 +21,8 @@ manifest:
   remotes:
     - name: upstream
       url-base: https://github.com/zephyrproject-rtos
+    - name: syst_upstream
```
Hmm shouldn't we be mirroring this at the project level, instead of adding a new upstream?

nashif replied:
this is still WIP I guess :)

wentongwu replied:
@andrewboie @nashif put the changes for syst on my own fork first, when everything is good, will do the mirroring at the project level.

wentongwu commented:
the output of failed test cases is like below which is what we wanted and also it can be decoded by the decoder, the CI failure may be caused by the wrong configure of the added sample, I will fix it.
 
```
SYS-T RAW DATA: 0200000B2A2A2A2A2A20426F6F74696E67205A6570687972204F53206275696C64207A65706879722D76322E302E302D313038382D67303264333035623465323134202A2A2A2A2A0A00
SYS-T RAW DATA: 2200000B4572726F72206D657373616765206578616D706C652E00
SYS-T RAW DATA: 3200000B5761726E696E67206D657373616765206578616D706C652E00
SYS-T RAW DATA: 4200000B496E666F206D657373616765206578616D706C652E00
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652E00
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652E00
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652C2025640001000000
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652C2025642C202564000100000002000000
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652C2025642C2025642C20256400010000000200000003000000
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652C2025642C2025642C2025642C2030782532780001000000020000000300000004000000
SYS-T RAW DATA: 2600001A03D2040000080102030405060708
SYS-T RAW DATA: 3600001A03D2040000080102030405060708
SYS-T RAW DATA: 4600001A03D2040000080102030405060708
SYS-T RAW DATA: 7600001A03D2040000080102030405060708
SYS-T RAW DATA: 0200000B68656C6C6F207379732D74206F6E20626F617264206D7073325F616E33383500
```

wentongwu commented:
@nashif could you please give some comments for this PR? Thanks

nashif commented on west.yml:
```diff
@@ -100,6 +102,10 @@ manifest:
     - name: littlefs
       path: modules/fs/littlefs
       revision: fe9572dd5a9fcf93a249daa4233012692bd2881d
+    - name: public-mipi-sys-t
```
please use github.com:zephyrproject-rtos/mipi-sys-t.git instead and put this in its own commit.

wentongwu replied:
ok

wentongwu replied:
@nashif have already used zephyrproject-rtos/mipi-sys-t.git.

nashif commented on include/mipi_syst/platform.h:
```diff
@@ -0,0 +1,106 @@
+/*
```
this is not a zephyr public API, you should move it under subsystem/logging or subsystem/debug.

wentongwu replied:
ok, Thanks

nashif commented on modules/Kconfig.syst:
```diff
@@ -0,0 +1,8 @@
+
+config MIPI_SYST_LIB
+	bool "MIPI SyS-T Library Support"
+	select NEWLIB_LIBC
```
why do you need newlib?
Also, move this to the same commit where you introduce the module in west.yml.

wentongwu replied:
sys-t lib requires some header files which can't be founded in Minimal libc.

nashif commented on subsys/logging/log_output_syst.c:
```diff
@@ -24,8 +24,8 @@
 
```
you probably want to fold this commit into the previous commit introducing this file.

nashif commented on samples/subsys/logging/syst/src/main.c:
```diff
@@ -14,7 +14,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_DBG);
 
```
squash into original commit introducing the change please

nashif commented on samples/subsys/logging/syst/sample.yaml:
```diff
@@ -4,3 +4,8 @@ tests:
   sample.subsys.logging.syst:
     platform_whitelist: mps2_an385
     tags: logging
+    harness: console
```
squash into original commit please

nashif commented on subsys/logging/CMakeLists.txt:
```diff
@@ -48,6 +47,16 @@ if(NOT CONFIG_LOG_MINIMAL)
     log_backend_qemu_x86_64.c
   )
+
+  zephyr_sources_ifdef(
```
That does not seems like the right approach here, @nordic-krch  can you please provide some feedback on how to support a different output back-end?

wentongwu replied:
@nashif IMHO back-end doesn't take care the output format, just pushing the data out with his own protocol(uart, usb, etc.). it's the place where in log_output.c logging subsystem decide the output format, so I replace the original one with sys-t format. I don't do the custom in the front-end, because I don't want to sys-t feature require more stack from the thread which use the front-end Marcos(LOG_INFO, LOG_ERR, etc).

@nordic-krch could you please help give your comments? Thanks.

wentongwu commented:
@nashif @nordic-krch merge window will close in 2 or 3 weeks, really appreciate you can give further comments about this PR. Thanks.

nordic-krch commented:
@wentongwu i don't like forking `log_output.c` into `log_output_syst.c`. there is a log of code redundancy between those two and it forces syst format on all backends. Isn't it possible to extend log_output will additional flags? Something similar to `LOG_OUTPUT_FLAG_FORMAT_SYSLOG`. This will allow to use syst format only for specific backend.

nordic-krch commented on modules/Kconfig.syst:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2019 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config MIPI_SYST_LIB
```
I would add option to the logger kconfig (like `LOG_MIPI_SYST_ENABLE`) which select this option. Otherwise there is no explicit connection between the logger and this library.

wentongwu replied:
Thanks,

nordic-krch commented on subsys/logging/log_backend_uart.c:
```diff
@@ -31,7 +31,12 @@ LOG_OUTPUT_DEFINE(log_output, char_out, &buf, 1);
-	log_backend_std_put(&log_output, 0, msg);
+	u32_t flag = 0;
+
+#ifdef CONFIG_LOG_BACKEND_UART_SYST_ENABLE
```
you could use `IS_ENABLED()`
something like
`u32_t flag = IS_ENABLED(SYST_ENABLE) ? LOG_OUTPUT_FLAG_FORMAT_SYST : 0;`
would look more concise, imo.

wentongwu replied:
Thanks, understand the IS_ENABLED

nordic-krch commented on subsys/logging/log_output.c:
```diff
@@ -506,6 +518,13 @@ void log_output_msg_process(const struct log_output *log_output,
 	int prefix_offset;
 
+#ifdef CONFIG_LOG_MIPI_SYST_ENABLE
+	if (flags & LOG_OUTPUT_FLAG_FORMAT_SYST) {
```
i cases like that (runtime + compile time condition) i prefer to create local function (e.g. `is_format_syst_enabled(flags)`) which simply wraps `IS_ENABLED(SYST_ENABLE) && (flags & SYST_FLAG)`. that reduces ifdefs to 0 (no need to ifdef function declarations too)

wentongwu replied:
Thanks

wentongwu commented:
> @wentongwu i don't like forking `log_output.c` into `log_output_syst.c`. there is a log of code redundancy between those two and it forces syst format on all backends. Isn't it possible to extend log_output will additional flags? Something similar to `LOG_OUTPUT_FLAG_FORMAT_SYSLOG`. This will allow to use syst format only for specific backend.

The initial idea is that I don't want to break the exiting log_output.c, but it needs the existing interfaces to have logging system work well, so I forked log_output.c into log_output_syst.c which implemented the needed APIs.

Now I restructure the code, put sys-t related code from log_output_syst.c to https://github.com/zephyrproject-rtos/mipi-sys-t and only leave the APIs in log_output_syst.c needed by log_output.c if enabled sys-t format for that back-end. Thanks for your suggestions, could you please take another look the new code? Thanks

nordic-krch commented on samples/subsys/logging/syst/src/main.c:
```diff
@@ -0,0 +1,90 @@
+	LOG_INF("Info message example.");
+	LOG_DBG("Debug message example.");
+
+	for (int i = 0; i <= 4; i++) {
```
what's the purpose of this for loop? why not simply put 5 log messages one by one?

andrewboie replied:
this confuses me too

wentongwu replied:
my fault

andrewboie commented on subsys/logging/log_backend_rtt.c:
```diff
@@ -221,7 +221,10 @@ LOG_OUTPUT_DEFINE(log_output, IS_ENABLED(CONFIG_LOG_BACKEND_RTT_MODE_BLOCK) ?
 		struct log_msg *msg)
 {
-	log_backend_std_put(&log_output, 0, msg);
+	u32_t flag = IS_ENABLED(CONFIG_LOG_BACKEND_RTT_SYST_ENABLE) ?
```
it doesn't seem good that we have to patch all the backends like this specifically for one output formatting type. it doesn't seem like this will scale well to additional format types as they are added. is there some other way we can do this in more centralized code?

nordic-krch replied:
alternative is enabling syst formatting for all backends at once. However, as i'm not familiar with syst i don't know what is the exact use case. How does it match with formatting used for net or bluetooth.

wentongwu replied:
currently I don't think there is a centralized way to enable syst formatting for all backends, because besides standard logger backends, there are non-standard backends like net backend or bluetooth backend.

wentongwu replied:
@nordic-krch @andrewboie @nashif besides centralized issue, is there any other concerns for this PR? Thanks

wentongwu committed:
logging: add syst format output

Add syst format output support for existing logging system.

Fixes: #19841.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
logging: add UART syst backend support

add UART syst backend support.

Fixes: #19841.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
logging: add SWO syst backend support

add SWO syst backend support.

Fixes: #19841.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
logging: add RTT syst backend support.

add RTT syst backend support.

Fixes: #19841.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
logging: add networking syst backend support.

add networking syst backend support.

Fixes: #19841.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
samples: add sample for logging syst format output

add sample for syst format output to prove the output can be
decoded by the existing decoder.

Fixes: #19841.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

nordic-krch commented on subsys/logging/log_output_syst.c:
```diff
@@ -0,0 +1,221 @@
+{
+	char buf[CONFIG_LOG_STRDUP_MAX_STRING + 1];
+	size_t length = CONFIG_LOG_STRDUP_MAX_STRING;
+	u32_t severity = level_to_syst_severity(log_msg_level_get(msg));
```
raw string has custom `LOG_LEVEL_INTERNAL_RAW_STRING` (5) here it will be converted to debug level. Not sure if that is an expected result. Raw string is used when printk goes through logger. Alternatively severity can be fixed here to 0.

wentongwu replied:
@nordic-krch yes, we use raw string if printk goes through logger, and also I have it tested in sample code like below.
```

        /* raw string */
        printk("hello sys-t on board %s\n", CONFIG_BOARD);
```
And checked code again, it seems `LOG_LEVEL_INTERNAL_RAW_STRING` already equals 0 as below marco defined. If anything wrong, please correct me. Thanks
`#define LOG_LEVEL_INTERNAL_RAW_STRING LOG_LEVEL_NONE`

nordic-krch replied:
my bad, somehow i thought it had different number.

nordic-krch commented:
PR looks good. I think it would be good to extend conceptual documentation to mention this feature Frankly speaking i'm not exactly sure how syst is going to be used.

nordic-krch reviewed (approved):
Hoping to see conceptual documentation extension in another PR.

jukkar commented on subsys/logging/Kconfig:
```diff
@@ -492,6 +492,12 @@ config LOG_BACKEND_NET_MAX_BUF_SIZE
 	  IPv6 the size is 1180 octets. As each buffer will use RAM, the value
 	  should be selected so that typical messages will fit the buffer.
 
+config LOG_BACKEND_NET_SYST_ENABLE
```
I was not marked as a reviewer for this PR so did not had a chance to comment this. I think this was merged a bit prematurely as now if user enables `CONFIG_LOG_BACKEND_NET` and `CONFIG_LOG_BACKEND_NET_SYST_ENABLE`, the loggers net backend prepares to send syslog formatted messages out but this will fail unless user sets the syslog server hostname. The code is prepared to handle that issue but there is now just useless code in the system if user only meant to activate syst logging and not syslog one.
Better would have been to have separate option for enabling syslog formatted network logging, just like the syst one.

wentongwu commented:
> Hoping to see conceptual documentation extension in another PR.

I will, thanks.
